runtime/lua/vim/lsp/buf.lua

```diff
@@ -76,8 +76,75 @@ end
 ---     - on_list: (function) |lsp-on-list-handler| replacing the default handler.
 ---       Called for any non-empty result.
 function M.definition(options)
+  options = options or {}
   local params = util.make_position_params()
-  request_with_options(ms.textDocument_definition, params, options)
+  local bufnr = api.nvim_get_current_buf()
+
+  ---@param results table<integer, {error: lsp.ResponseError, result: lsp.Location|lsp.Location[]|lsp.LocationLink[]|nil}>
+  local function handler(results)
+    local client_id_to_items = {}
+    local total_items = 0
+    for client_id, v in pairs(results) do
+      if v.error == nil and v.result ~= nil then
+        -- first check for Location | LocationLink because
+        -- tbl_islist returns `false` for empty lists
+        if v.result.uri or v.result.targetUri then
+          local client = vim.lsp.get_client_by_id(client_id)
+          local items = util.locations_to_items({ v.result }, client.offset_encoding)
+          total_items = total_items + #items
+          client_id_to_items[client_id] = items
+        elseif vim.tbl_islist(v.result) then
+          local client = vim.lsp.get_client_by_id(client_id)
+          local items = util.locations_to_items(v.result, client.offset_encoding)
+          total_items = total_items + #items
+          client_id_to_items[client_id] = items
+        end
+      end
+    end
+
+    if total_items == 0 then
+      return
+    end
+
+    -- jump to the only response
+    if total_items == 1 then
```

Comment:
We should still check if `options.on_list` is given here and invoke it if so. See https://github.com/neovim/neovim/pull/25830 for context.

Reply:
Thanks, fixed in [4e72d41](https://github.com/neovim/neovim/pull/26507/commits/4e72d41ca5833f174c340d19fdefc13f4e4f91c8)